lib/withOnyx.d.ts

```diff
@@ -0,0 +1,27 @@
+import {IsEqual} from 'type-fest';
+import {DeepKeyOf, TypeOptions} from './types';
+
+type Key = TypeOptions['keys'];
+type Value = TypeOptions['values'];
+
+type Mapping<TComponentProps, TOnyxProps, TMappingKey extends keyof TOnyxProps, TOnyxKey extends Key> = IsEqual<Value[TOnyxKey] | null, TOnyxProps[TMappingKey]> extends true
+    ? {
+          // key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey); // FIXME: Breaks the key/value inference.
+          key: TOnyxKey;
+          canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+          initWithStoredValues?: boolean;
+          selector?: Value[TOnyxKey] extends object | string | number | boolean ? ((value: Value[TOnyxKey] | null) => TOnyxProps[TMappingKey]) | DeepKeyOf<Value[TOnyxKey]> : never;
+      }
+    : never;
+
+type KeyValueMapping<TComponentProps, TOnyxProps, TMappingKey extends keyof TOnyxProps> = {
```

Comment:
@hayata-suenaga @blazejkustra Coming from [here](https://github.com/Expensify/react-native-onyx/pull/267#issuecomment-1632060418).

This "magic" type allows me to keep type inference separately for each property in the mapping object.